### PR TITLE
The Generative drop v0

### LINKED ISSUE #7580: The Generative drop v0
### Context 

In KodaDot we are huge fans of generative and programatic art in KodaDot.

###   From UI perspective 

1. we first need to get drop of type `generative` waiting for #7570

2.   By default image of the art would be shown

3.  When user press run button `image` would change to `iframe` where `src` would be `content` + ?`hash={hash}  `.  **by default hash = accountId**

variations 00

￼
4.   When user clicks variation you create new `hash = encodeAddress(accountId, newss58)`
 where newss58 is number from 0 up to 15000 https://github.com/paritytech/ss58-registry/blob/30889d6c9d332953a6e3333b30513eef89003f64/ss58-registry.json#L1292C17-L1292C22 
 
 > at this step it is highly possible that you will re-mount `iframe` 

5.   When user is satisfied with the iteration he or she creates and iteration 

6.   Metadata are constructed on frontend as follows:

```yml
 name: same as collection
 description: same as collection 
animation_url: content + hash
image: same as collection, wating for https://github.com/kodadot/workers/issues/169
iterationHash: hash
__other to be added
```

### PR DESCRIPTION
**Pull request by Jarsen136**
**Thank you for your contribution** to the [KodaDot - One Stop Shop for Polkadot NFTs](https://kodadot.xyz).

  👇 __ Let's make a quick check before the contribution.

  ## PR Type

  - [x] Feature

  ## Needs QA check

  - @kodadot/qa-guild please review

  ## Context

  - [x] Closes #7580
  - [ ] Requires deployment <snek/rubick/worker>

  #### Did your issue had any of the "$" label on it?

  - [x] My DOT address: [Payout](https://canary.kodadot.xyz/dot/transfer/?target=16SjUbGKSdjCdWTy3NNT3JxbRVGGqD4mwkHpc6BD9U2Rp29Z)

  ## Screenshot 📸

  - [ ] My fix has changed UI

  ## Copilot Summary
  
### <samp>🤖 Generated by Copilot at 7e44249</samp>

This pull request adds support for generative NFTs with unlockable content of different MIME types. It introduces a new component `Generative.vue` and several subcomponents to render the generative drop UI. It also updates the `dropCollectionById` query, the `useDrops` hook, and the `pages/[prefix]/drops/[id].vue` file to handle the new drop type. It adds a new function `generativeTokenMintArgs` and modifies the `createUnlockableMetadata` function to create the mint transaction arguments and metadata hash. It also adds some translation keys to the `locales/en.json` file.

  
### <samp>🤖 Generated by Copilot at 7e44249</samp>

> _Oh we are the coders of the NFT sea_
> _We make the drops generative and unlockable_
> _We use the hooks and the components with `useGraphql`_
> _And we mint the tokens with `generativeTokenMintArgs`_

### CODE REVIEW

**Comment by Jarsen136** [comment]
> when meta > 0  you need to call api.tx.nfts.mint(collectionId, collection.count, accountId)

I got stuck at the final step of minting NFT. When I try to send the tx, it fails and tells me I have no permission to operate this collection. Would you mind helping me with this PR and pointing out what's wrong with it? @vikiival

**Comment by vikiival** [comment]
> pointing out what's wrong with it? @vikiival 
> 
Please follow 

https://github.com/kodadot/nft-gallery/issues/7580#issuecomment-1768488000

**Comment by vikiival** [comment]
> I did follow the guide. The meta of the generative drop is 1000000000

Meta should be null in this case i fixed the DB

**Comment by Jarsen136** [comment]
> > I did follow the guide. The meta of the generative drop is 1000000000
> 
> Meta should be null in this case i fixed the DB

I have updated them accordingly. 
Which API should I call? I have tried with api `/do/genchains` and it responded successfully but there is no nft created then.

**Comment by vikiival** [comment]
> successfully but there is no nft created then.

Because it failed https://assethub-kusama.subscan.io/extrinsic/0xf4f157a609689979848dcd87c68dade6a996e96953ebd3433296353f5b92fffd

**Comment by Jarsen136** [comment]
> > successfully but there is no nft created then.
> 
> Because it failed https://assethub-kusama.subscan.io/extrinsic/0xf4f157a609689979848dcd87c68dade6a996e96953ebd3433296353f5b92fffd


As this transaction was emitted from the server side, what should we do to deal with this failure?
The only difference I noticed is that `mint_to` points to another address instead of the collection's owner. 
image

**Comment by vikiival** [comment]
It should point to your address, honestly i do not know why it failed

**Comment by Jarsen136** [comment]
> It should point to your address, honestly i do not know why it failed


---

**Only the collection's owner has permission to mint nft and send it to other address. I bet that's why the transaction failed. WDYT? @vikiival** 


---

For the previous successful transaction, it was emitted by the collection's owner. 

emitted by: Gn84LKb5HSxc3SACayxCzKQcWESRMcT1VUCqeZURfGj6ASi
collection id: 165 
collection's owner: Gn84LKb5HSxc3SACayxCzKQcWESRMcT1VUCqeZURfGj6ASi

image
image


However, for the failed one, it was emitted by the  `Gn84xxx6ASi` instead of collection's owner (vikiival)
emitted by: Gn84LKb5HSxc3SACayxCzKQcWESRMcT1VUCqeZURfGj6ASi
collection id: 176 
collection's owner: vikiival

image

**Comment by vikiival** [comment]
> Only the collection's owner has permission to mint nft and send it to other address. I bet that's why the transaction failed.

Yup I had really weird settings on this collection (according to blockchain - KodaBot was the owner, but vikiival had all the rights to mint.

TL;DR I was able to mint but

This timer is too low. I waited around a mintute to be redirected. So please show a correct time.

Screenshot 2023-10-23 at 12 39 27



2. When I was playing with settings I broke the squid. The fix should be propagated in 10 min or less

- https://github.com/kodadot/stick/pull/130

**Comment by Jarsen136** [comment]
> Yup I had really weird settings on this collection (according to blockchain - KodaBot was the owner, but vikiival had all the rights to mint.
> 
> TL;DR I was able to mint but
> 
Interesting : ) Now It works for me as well.

> This timer is too low. I waited around a mintute to be redirected. So please show a correct time.
> 
Updated

File: components/collection/unlockable/utils.ts [hunk 1]
```diff
@@ -20,15 +20,17 @@ export async function createUnlockableMetadata(
...
   const meta = createMetadata(
     name,
     description,
     imageHash,
-    undefined,
+    animationUrl,
     [],
     'kodadot.xyz',
-    'image/png',
+    miniType,
```
Comment (vikiival) [inline components/collection/unlockable/utils.ts:33]: mimetype

File: components/collection/drop/GenerativePreview.vue [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  content: string
+  image?: string
+}>()
+
+const emit = defineEmits(['select'])
+
+const { accountId } = useAuth()
+const getHash = () => {
+  // https://github.com/paritytech/ss58-registry/blob/30889d6c9d332953a6e3333b30513eef89003f64/ss58-registry.json#L1292C17-L1292C22
+  return encodeAddress(accountId.value, getRandomInt(15000))
```
Comment (vikiival) [inline components/collection/drop/GenerativePreview.vue:45]: missing to hex

**Comment by vikiival** [comment]
1. you are missing trailing slash

Now
```
ipfs://bafybeigsw7gagsmvxxivt5kvrl6ueld7yszzef2aylxbzzafez6ybxscca?hash=nkWZwxT9BDJsC58WaPW9hyCmqyC7r6NCece4aQnJYSNtguWTM
```

Expected (mind the `/` before `?`)

```
ipfs://bafybeigsw7gagsmvxxivt5kvrl6ueld7yszzef2aylxbzzafez6ybxscca/?hash=nkWZwxT9BDJsC58WaPW9hyCmqyC7r6NCece4aQnJYSNtguWTM
```

2. NFT view is broken

https://image.w.kodadot.xyz/ipfs/bafybeigsw7gagsmvxxivt5kvrl6ueld7yszzef2aylxbzzafez6ybxscca/?hash=nkWZwxT9BDJsC58WaPW9hyCmqyC7r6NCece4aQnJYSNtguWTM

Screenshot 2023-10-23 at 15 50 01

Vs view


Screenshot 2023-10-23 at 15 51 18

File: components/collection/unlockable/utils.ts (hunk 1 above)
Comment (Jarsen136) [reply components/collection/unlockable/utils.ts]: Done

File: components/collection/drop/GenerativePreview.vue (hunk 2 above)
Comment (Jarsen136) [reply components/collection/drop/GenerativePreview.vue]: done

**Comment by Jarsen136** [comment]
> 1. you are missing trailing slash

> ```
> 
> 2. NFT view is broken

> 3. In GenerativePreview.vue L:45 should be

✅ All of them have been fixed. Thank you for pointing them out with the code.

**Comment by Jarsen136** [comment]
> Some questions: Mint button is supposed to work only after clicking on Run? is this normal behavior? 

Yes, the user should click the button and choose their favourite image.


> Variations sometimes take a long time to load here for me(5s), maybe a small loading icon could tell the user that the variation is loading?
> 
Updated. I have added the loading icon for it.

> creation goes nicely: https://deploy-preview-7745--koda-canary.netlify.app/ahk/gallery/176-4
> 
> recent NFTs mint still needs implementation: image

Done

image

**Comment by Jarsen136** [comment]
> > Yes, the user should click the button and choose their favourite image.
> 
> I would default it with ss58 of current chain ()

✅  Updated. I set a default value for it so the user does not have to click the `run` button.

**Comment by prury** [comment]
one last thing, if i understood correctly, the image I'm seeing on the right side is what i get when i click mint, so, instead of example of drop we could change it to something else, maybe "Select your Variation" and change the Variation button to Generate. @exezbcz wdyt? 

image

**Comment by Jarsen136** [comment]
> > Done
> > image
> 
> they all have the same cover image even if NFTs are different

It's intended because we set the same preview image for every image.  
ref https://github.com/kodadot/nft-gallery/issues/7580#issue-1935121606

image

**Review by vikiival — APPROVED** [review]

**Comment by vikiival** [comment]
@prury / @Jarsen136 Is there any blocker for this PR

**Comment by vikiival** [comment]
Also can someone from @kodadot/internal-dev review?

**Comment by Jarsen136** [comment]
> @Jarsen136 Is there any blocker for this PR?

No. It's ready to review and merged.

File: components/collection/drop/GenerativePreview.vue [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+const getHash = (isDefault?: boolean) => {
+  const ss58Format = isDefault
+    ? chainProperties.value?.ss58Format
+    : getRandomInt(15000)
+  // https://github.com/paritytech/ss58-registry/blob/30889d6c9d332953a6e3333b30513eef89003f64/ss58-registry.json#L1292C17-L1292C22
+  return stringToHex(encodeAddress(accountId.value, ss58Format))
+}
+
+const generativeImageUrl = ref(
```
Comment (roiLeo) [inline components/collection/drop/GenerativePreview.vue:52]: computed maybe?

File: components/collection/drop/GenerativePreview.vue [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    isLoading.value = false
+  }, 3000)
+}
+</script>
+
+<style scoped lang="scss">
+@import '@/assets/styles/abstracts/variables';
+
+.fixed-size {
+  width: 580px;
```
Comment (roiLeo) [inline components/collection/drop/GenerativePreview.vue:80]: ☹️

**Review by roiLeo — APPROVED** [review]
otherwise lgtm

**Comment by prury** [comment]
> @prury / @Jarsen136 Is there any blocker for this PR?

was just waiting about Exez opinion on this https://github.com/kodadot/nft-gallery/pull/7745#issuecomment-1777142078, but will open a followup instead

**Comment by exezbcz** [comment]
@prury I would remove the tag and create followup for more UI changes

**Comment by prury** [comment]
> @prury I would remove the tag and create followup for more UI changes

@Jarsen136 please remove:
image
and we are good to go

**Comment by yangwao** [comment]
I think there will be few refactor/iterations to improve codebase, now we can at least test it 

Thanks!
pay 50 usd

**Comment by yangwao** [comment]
😍 Perfect, I’ve sent the payout 
💵 $50 @ 4.34 USD/DOT ~ 11.521 $DOT
🧗 [16SjUbGKSdjCdWTy3NNT3JxbRVGGqD4mwkHpc6BD9U2Rp29Z](https://polkadot.subscan.io/account/16SjUbGKSdjCdWTy3NNT3JxbRVGGqD4mwkHpc6BD9U2Rp29Z)
🔗 [0x25170e0a1c3d08cde8726caad0cfdc533453b10fe1367626b7ecef058ec0a55e](https://polkadot.subscan.io/extrinsic/0x25170e0a1c3d08cde8726caad0cfdc533453b10fe1367626b7ecef058ec0a55e)

🪅 Let’s grab another issue and get rewarded! 
🪄 [github.com/kodadot/nft-gallery/issues](https://github.com/kodadot/nft-gallery/issues/)

File: components/collection/drop/GenerativePreview.vue (hunk 3 above)
Comment (Jarsen136) [reply components/collection/drop/GenerativePreview.vue:52]: No, it's only an initial value

File: components/collection/drop/Generative.vue [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+  () => collectionData.value?.collectionEntity?.max || 200,
+)
+const totalAvailableMintCount = computed(
+  () => totalCount.value - mintedCount.value,
+)
+
+watch(accountId, () => {
+  tryAgain({
+    account: accountId.value,
+  })
```
[the remark this answers is not in the document]
Comment (Jarsen136) [reply components/collection/drop/Generative.vue:175]: I learned it! Thank you

File: components/collection/drop/GenerativePreview.vue (hunk 4 above)
Comment (Jarsen136) [reply components/collection/drop/GenerativePreview.vue:80]: Updated on another pr

**Pull request merged**